Source/IGListGridCollectionViewLayout.m

```diff
@@ -0,0 +1,422 @@
+    self.lineNumber = (self.collectionView.numberOfSections + self.itemPerLine - 1) / self.itemPerLine;
+    self.interitemSpacing = (self.contentWidth - (self.itemSize.width * self.itemPerLine)) / (self.itemPerLine - 1);
+}
+
+@end
+
+#pragma mark - _IGListGridLayoutLine Implementation
+
+@implementation _IGListGridLayoutLine
+
+- (id)initWithMinimumInteritemSpacing:(CGFloat)spacing
+                            headIndex:(NSInteger)headIndex
+                                frame:(CGRect)frame
+                      scrollDirection:(UICollectionViewScrollDirection)direction
+
+{
+    self = [super init];
+    if (self) {
+        _frame = frame;
+        _scrollDirection = direction;
+        _minimumInteritemSpacing = spacing;
+        _itemSizes = [NSMutableArray array];
+        _headIndex = headIndex;
+        _tailSpace = frame.size.width - self.minimumInteritemSpacing;
+    }
+    return self;
+}
+
+- (BOOL)addItemToTailWithSize:(CGSize)size
+{
+    if (size.width > self.tailSpace) {
+        return NO;
+    }
+    
+    self.tailSpace -= size.width + self.minimumInteritemSpacing;
+    if (size.height > self.frame.size.height) {
+        CGRect frame = self.frame;
+        frame.size.height = size.height;
+        self.frame = frame;
+    }
+    NSValue *sizeValue = [NSValue valueWithCGSize:size];
+    [self.itemSizes addObject:sizeValue];
+    return YES;
+}
+
+- (UICollectionViewLayoutAttributes *)attributesForItemAtIndexPath:(NSIndexPath *)indexPath
+{
+    NSInteger index = indexPath.section - self.headIndex;
+    CGFloat x = 0;
+    NSInteger idx = 0;
+    for (NSValue *sizeValue in self.itemSizes) {
+        if (idx < index) {
+            CGSize size = [sizeValue CGSizeValue];
+            x += size.width + self.minimumInteritemSpacing;
+        } else {
+            break;
+        }
+        idx++;
+    }
+    UICollectionViewLayoutAttributes *attributes = [self attributesForItemAtIndexPath:indexPath withXOffset:x];
+    return attributes;
+}
+
+- (NSArray<UICollectionViewLayoutAttributes *> *)attributesForAllItems
+{
+    NSMutableArray *array = [NSMutableArray array];
+    CGFloat x = 0;
+    NSInteger idx = 0;
+    for (NSValue *sizeValue in self.itemSizes) {
+        NSIndexPath *indexPath = [NSIndexPath indexPathForItem:0 inSection:(self.headIndex + idx)];
+        UICollectionViewLayoutAttributes *attributes = [self attributesForItemAtIndexPath:indexPath withXOffset:x];
+        [array addObject:attributes];
+        CGSize size = [sizeValue CGSizeValue];
+        x += size.width + self.minimumInteritemSpacing;
+        idx++;
+    }
+    return array;
+}
+
+#pragma mark - Private API
+
+- (UICollectionViewLayoutAttributes *)attributesForItemAtIndexPath:(NSIndexPath *)indexPath withXOffset:(CGFloat)x
+{
+    NSInteger index = indexPath.section - self.headIndex;
+    CGSize itemSize = [self.itemSizes[index] CGSizeValue];
+    
+    // Center vertically
+    CGFloat y = (self.frame.size.height - itemSize.height) / 2;
```

Comment:
This division by int `2` is interesting, not sure if rounding is intended or if we should divide by `2.0`